File: amazon-interview/recursion/staircase.rb
  def stepPerms(n)
    return 1 if n == 1 || n == 0

    steps = [1,2,3]
    cache = {}

    total = 0
    steps.each do |step|
      if n - step == 0
        total += 1
      elsif n - step > 0
        if cache[n-step].nil?
          cache[n-step] = stepPerms(n - step)
        end
        total += cache[n-step]
      end
    end
    total
  end

File: amazon-interview/recursion/staircase.py

def stepPerms(n):
    steps = [1, 2, 3]
    ways = dict()

    def climb(n, steps, ways):
        ret = 0
        for step in steps:
          if n - step == 0:
              ret += 1
          elsif n - step > 0:
              if n - step not in ways:
                  ways[n - step] = climb(n - step, steps, ways)
              print(ways[n - step])
              ret += ways[n - step]
              return ret

    return climb(n, steps, ways)

stepPerms(4)

